linera-explorer/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+        .expect("failed to serialize ChainId");
+    setf(app, "chain", &chain_js);
+    let path = match (path, data.page.clone()) {
+        (Some(p), _) => p,
+        (_, Page::Unloaded | Page::Home { .. }) => "",
+        (_, Page::Block(_)) => "block",
+        (_, Page::Blocks { .. }) => "blocks",
+        (_, Page::Applications(_)) => "applications",
+        (_, Page::Error(_)) => "error",
+    };
+    let address = node_service_address(&data.config, false);
+    let result = match path {
+        "" => home(&address, chain_id).await,
+        "block" => {
+            let hash = args.iter().find_map(|(k, v)| {
+                if k == "block" {
+                    Some(CryptoHash::from_str(v).unwrap())
+                } else {
+                    None
+                }
+            });
+            block(&address, chain_id, hash).await
+        }
+        "blocks" => blocks(&address, chain_id, None, Some(20)).await,
+        "applications" => applications(&address, chain_id).await,
+        "error" => {
+            let msg = args
+                .iter()
+                .find_map(|(k, v)| {
+                    if k == "msg" {
+                        Some(v.to_string())
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or("unknown error".to_string());
+            Err(msg)
+        }
+        _ => Err("unknown page".to_string()),
+    };
+    let (page, new_path) = result.unwrap_or_else(|e| error(&e));
+    let page_js = format_bytes(&page.serialize(&SER).unwrap());
+    setf(app, "page", &page_js);
+    let new_path = format!("{}?chain={}", new_path, chain_id);
+    web_sys::window()
+        .expect("window object not found")
+        .history()
+        .expect("history object not found")
+        .push_state_with_url(&page_js, &new_path, Some(&new_path))
+        .expect("push_state failed");
+}
+
+#[wasm_bindgen]
+pub async fn route(app: JsValue, path: JsValue, args: JsValue) {
+    let path = path.as_string();
+    let args = from_value::<Vec<(String, String)>>(args).unwrap_or(vec![]);
+    let msg = format!(
+        "route: {} {:?}",
+        path.clone().unwrap_or("none".to_string()),
+        args
+    );
+    log_str(&msg);
+    let data = from_value::<Data>(app.clone()).expect("cannot parse vue data");
+    let chain_id = args
+        .iter()
+        .find_map(|(k, v)| {
+            if k == "chain" {
+                Some(ChainId::from_str(v).expect("wrong chain id"))
+            } else {
+                None
+            }
+        })
+        .unwrap_or(data.chain);
+    route_aux(&app, &data, &path, &chain_id, &args).await
+}
+
+#[wasm_bindgen]
+pub fn short(s: String) -> String {
+    let n = s.len();
+    format!("{}..{}", &s[..4], &s[n - 4..])
+}
+
+#[wasm_bindgen]
+pub fn short_app(s: String) -> String {
+    format!("{}..{}..{}..", &s[..4], &s[64..68], &s[152..156])
+}
+
+fn set_onpopstate(app: JsValue) {
```

Comment:
This could also use a doc string.

This is if the user clicks "back"?

Does Vue automatically take care of pushing the appropriate state, so that `v.state` contains it?

Should `v` be called `event`?

Reply:
no, it is not autmoatic, there is a `history.pushState` at the end of `route_aux` function